[Week1] Todo List 만들기

goodcodings opened the pull request:
fork 후 머지리퀘스트 하는게 아직 적응되지 않아 너무 급한 마음에 올리네요 ㅠㅠ

goodcodings committed:
task 도메인생성

Signed-off-by: Lee Byoung Seong <dev.bslee@gmail.com>

goodcodings committed:
[Week1] 빌더패턴 제거

goodcodings committed:
[Week1] 빌더패턴 제거

goodcodings commented:
안녕하세요 어제 야살님께서 **ParameterObject** 을 말씀하셔서 구글링을 해서 찾아봤음에도 개념이해가 가질 않아서 질문을 드립니다...
ParameterObject에 대해서 찾아봤을때  쓰임새가 같은 변수끼리 묶어서 객체를 생성한다라고 저는 이해했는데....
혹시 제가 맞게 이해한건지... 궁금해서 여쭤봅니다...

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/domain/TaskInfo.java:
```diff
@@ -0,0 +1,53 @@
+        private final LocalDateTime createdAt;
+        private final LocalDateTime updatedAt;
+
+        public TaskInfoBuilder(String taskId, String title, LocalDateTime createdAt, LocalDateTime updatedAt) {
```
안녕하세요! 빌더패턴을 사용하고 싶으신거라고  생각되는데요!
제가 아는 빌더패턴은 두가지가 있어요!

1. 객체를 생성할 때 인자가 많아 생성할 때 실수가 잦을 것 같은 경우 각 인자에 이름을 붙일 수 있도록 하는 것
2. 객체를 생성할 때 인자가 많고 생성방식이 복잡하여, 객체생성하긴 위한 순서을 지시하는 Director와 그 생성방법을 가지는 Builder를 분리하는 것

다음 코드에서는 아직 생성방식이 복잡한 것도 아니고 `TaskInfoBuilder` 안에 `build` 라는 메서드가 있는 것으로 보아 1번의 방식의 빌더패턴이라고 저는 가정해보겠습니다.
`build` 라는 메서드는 정확하게 동작하겠지만 아직 아쉬운 점이 있습니다! 

바로 그것은 `TaskInfoBuilder` 의 생성자입니다. 이 생성자는 4개를 가지지만 `String` 이 2개이며, `LocalDateTime` 이 둘은 각각 같은 타입이기 때문에 생성을 할 때 실수를 할 수 있어요! 예를 들어볼게요!
```java
new TaskInfo.TaskInfoBuilder("title-1", "title-2", ,,,);
```

동일한 String이기 때문에 이 둘이 유효한지는 확인하기 쉽지 않습니다! 
그렇기 때문에 2가지 방법으로 이를 해결할 수 있을 것 같아요.

1. 각 인자에 의미가 맞는 도메인 객체를 만든다. (ex. `TaskId` , `Title`)
2. 혹은 생성자를 이용한 방법이 아닌 빌더패턴을 이용하여 하나씩 인자를 받도록한다. (예를들어 아래와 같은 형태입니다)

  ```java
    new TaskInfo.TaskInfoBuilder()
                          .id(taskId)
                          .title(title)
                          .createdAt(createdAt)
                          .updatedAt(updatedAt)
                          .build();
  ```
이런 형태일 수 있을 것 같아요!

또한 어떤 의도로 만드셨는지 알려주시면 리뷰에 더 큰 도움을 드릴 수 있을 것 같아요!

goodcodings replied:
네! 감사합니다.
해당 객체는 서버에서 클라이언트 단에 응답을 해줄때 데이터 형식???의 역할을 위임하기 위해 만든 객체 입니다!!
생성자가 퍼블릭이란걸 제가 생각을 못했네요 ㅠㅠ

Task라는 객체는 디스크 기반 데이터베이스에 접근하기 위한 데이터 형식이라는걸 역할을 위임하는 객체라는걸 가정하고 만들었습니다!

goodcodings committed:
[Week1] Todo REST api 만들기
1. DIP를 최대한 지키기 위해 역할과 나눴습니다
   - 의존 주입을하는 AppConfig 클래스를 생성했습니다.
2. 에릭 에반스의 ddd sample을 참조하여 패키지 구조를 잡았습니다.
3. task의 빌더 패턴을 다시 만들었고 title 변경시 task 객체의 자율성을 위해 updateTitle 메소드를  생성했습니다

Lee-Byoung-Seong committed:
[Week1] Todo REST api 만들기
1. DIP를 최대한 지키기 위해 역할과 구현을 나눴습니다
- 의존 주입을하는 AppConfig 클래스를 생성했습니다.
2. 에릭 에반스의 ddd sample을 참조하여 패키지 구조를 잡았습니다.
3. task의 빌더 패턴을 다시 만들었고 title 변경시 task 객체의 자율성을 위해 updateTitle 메소드를 생성했습니다
4. httpHandler를 구현하여 클라이언트로 부터 요청 받고 응답하는 코드를 생성했습니다
5. Todo 생성까지 구현했습니다
6. 각 메소드에 맞는 처리를 하기 위해 post,get,put,delete클래스를 생성했습니다.

Lee-Byoung-Seong committed:
[Week1] Todo REST api 만들기
1. DIP를 최대한 지키기 위해 역할과 구현을 나눴습니다
- 의존 주입을하는 AppConfig 클래스를 생성했습니다.
2. 에릭 에반스의 ddd sample을 참조하여 패키지 구조를 잡았습니다.
3. task의 빌더 패턴을 다시 만들었고 title 변경시 task 객체의 자율성을 위해 updateTitle 메소드를 생성했습니다
4. httpHandler를 구현하여 클라이언트로 부터 요청 받고 응답하는 코드를 생성했습니다
5. Todo 생성까지 구현했습니다
6. 각 메소드에 맞는 처리를 하기 위해 post,get,put,delete클래스를 생성했습니다.

Lee-Byoung-Seong committed:
[Week1] Todo REST api 만들기
1. DIP를 최대한 지키기 위해 역할과 구현을 나눴습니다
- 의존 주입을하는 AppConfig 클래스를 생성했습니다.
2. 에릭 에반스의 ddd sample을 참조하여 패키지 구조를 잡았습니다.
3. task의 빌더 패턴을 다시 만들었고 title 변경시 task 객체의 자율성을 위해 updateTitle 메소드를 생성했습니다
4. httpHandler를 구현하여 클라이언트로 부터 요청 받고 응답하는 코드를 생성했습니다
5. Todo 생성까지 구현했습니다
6. 각 메소드에 맞는 처리를 하기 위해 post,get,put,delete클래스를 생성했습니다.
